toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1313,6 +1346,54 @@ func (g *PkgGraph) fixIntraSpecCycle(trimmedCycle []*PkgNode) (err error) {
 	return
 }
 
+// fixPrebuiltSRPMsCycle attempts to fix a cycle if at least one node is a pre-built SRPM.
+// If a cycle can be fixed, edges representing the build dependencies of the pre-built SRPM will be removed.
+func (g *PkgGraph) fixPrebuiltSRPMsCyclewithPMC(trimmedCycle []*PkgNode, outDir, tmpDir, workerTar, existingRpmsDir, existingToolchainRpmsDir string, usePreviewRepo bool, repoFiles []string, disableUpstreamRepos bool, tlsClientKey string, tlsClientCert string) (err error) {
+	logger.Log.Debug("Checking if cycle contains pre-built SRPMs.")
```

Comment:
It looks like we're repeating a lot of code from `fixPrebuiltSRPMsCycle`. Please refactor the code to extract the common part to avoid code repetition.

Reply:
I feel like the main difference is the function setting the `isPrebuilt` variable. If that's the case, I'd suggest merging the two functions and conditionally call `IsSRPMAvailablePMC`, if `IsSRPMPrebuilt` doesn't find a pre-built SRPM and we allow using PMC for cycle resolutions.

Reply:
yes, what the functions do is similar.
- fixPrebuiltSRPMsCycle checks local RPMs folder and replaces the node with a prebuilt node
- fixPrebuiltSRPMsCyclewithUpstream checks upstream-repos and replaces the node with remote-unresolved node (to be fetched by graphpkgfetcher)

These functions are not merged because:
1. IsSRPMAvailableUpstream requires workerTar, repofiles and other args which are not required by the IsSRPMPrebuilt function. These arguments have to be passed from grapher through makeDAG. A new makeDAG is created to avoid bloating the function signature of makeDAG and all the following functions with multiple parameters.

2. makeDAG is also used at the scheduler stage.

makeDAGusingUpstream marks the node as unresolved, assuming that the graphpkgfetcher stage that follows will fetch the RPM, which is not the case with scheduler stage.


Existing flow:
makeDAG -> fixCycle -> fixPrebuiltSRPMsCycle -> IsSRPMPrebuilt -> findAllRPMS

New flow:
makeDAGusingUpstream -> fixPrebuiltSRPMsCycleUsingUpstream -> IsSRPMAvailableUpstream -> findAllRPMSUpstream -> **cloner** (Initialize cloner requires multiple args)

I can merge both the flows as follows, by modify the existing makeDAG signature, if this seems better. @PawelWMS 
makeDAG -> fixCycle -> fixPrebuiltSRPMsCycle, fixPrebuiltSRPMsCycleUsingUpstream

(Trying to resolve cycle with local RPMs first and if not found, then only look in upstream repos)